File: R/access.R
#'
#'    access to internal functions, for debugging
#'
#'   kermom()  should agree with kernel.moment()
#' 
#'  Copyright (c) 2023 Adrian Baddeley, Tilman Davies and Martin Hazelton
#'  GNU Public Licence (>= 2.0)

kermom <- function(m, r, kernel="gaussian", mean = 0,
                   sd=1/kernel.factor(kernel)) {
  kernel <- match.kernel(kernel)
  kerncode <- switch(kernel,
                     gaussian=1,
                     rectangular=2,
                     triangular=3,
                     epanechnikov=4,
                     biweight=5,
                     cosine=6,
                     optcosine=7,
                     0)
  df <- data.frame(r=r, mean=mean, sd=sd)
  nr <- nrow(df)
  z <- .C(SK_kermom,
          nx = as.integer(nr),
          x = as.double(df$r),
          mean = as.double(df$mean),
          sd = as.double(df$sd),
          m = as.integer(m),
          kerncode = as.integer(kerncode),
          y = as.double(numeric(nr)),
          errcode = as.integer(integer(1)),
          PACKAGE="spatstat.univar")
  if(z$errcode != 0)
    switch(z$errcode,
           stop("Error in C: negative length"),
           stop("Error in C: unrecognised kernel"),
           stop("Error in C: value of m is not supported"),
           stop(paste("Error in C: error code", z$errcode)))
  return(z$y)
}


          

File: src/access.c
/*
  access.c

  R interface functions
  just to check the validity of internal code

*/

#include "interfacecodes.h"
#include "kernels.h"

void kermom(int *nx,     
	    double *x,
	    double *mean,
	    double *sd,
	    int *m, 
	    int *kerncode,
	    double *y,
	    int *errcode) {
  int i, n;
  n = *nx;
  *errcode = OK;
  
  if(*m == 0) {
    /* cumulative distribution function */
    if(*kerncode == GAUSSIAN) {
      for(i = 0; i < n; i++) y[i] = pgaussian(x[i], mean[i], sd[i]);
    } else if(*kerncode == RECTANGULAR){ 
      for(i = 0; i < n; i++) y[i] = prectangular(x[i], mean[i], sd[i]);
    } else if(*kerncode == TRIANGULAR) {
      for(i = 0; i < n; i++) y[i] = ptriangular(x[i], mean[i], sd[i]);
    } else if(*kerncode == EPANECHNIKOV) {
      for(i = 0; i < n; i++) y[i] = pepanechnikov(x[i], mean[i], sd[i]);
    } else if(*kerncode == BIWEIGHT) {     
      for(i = 0; i < n; i++) y[i] = pbiweight(x[i], mean[i], sd[i]);
    } else if(*kerncode == COSINE) {       
      for(i = 0; i < n; i++) y[i] = pcosine(x[i], mean[i], sd[i]);
    } else if(*kerncode == OPTCOSINE) {
      for(i = 0; i < n; i++) y[i] = poptcosine(x[i], mean[i], sd[i]);
    }
  } else if(*m == 1) {
    /* partial first moment */
    if(*kerncode == GAUSSIAN) {
      for(i = 0; i < n; i++) y[i] = m1gaussian(x[i], mean[i], sd[i]);
    } else if(*kerncode == RECTANGULAR){ 
      for(i = 0; i < n; i++) y[i] = m1rectangular(x[i], mean[i], sd[i]);
    } else if(*kerncode == TRIANGULAR) {
      for(i = 0; i < n; i++) y[i] = m1triangular(x[i], mean[i], sd[i]);
    } else if(*kerncode == EPANECHNIKOV) {
      for(i = 0; i < n; i++) y[i] = m1epanechnikov(x[i], mean[i], sd[i]);
    } else if(*kerncode == BIWEIGHT) {     
      for(i = 0; i < n; i++) y[i] = m1biweight(x[i], mean[i], sd[i]);
    } else if(*kerncode == COSINE) {       
      for(i = 0; i < n; i++) y[i] = m1cosine(x[i], mean[i], sd[i]);
    } else if(*kerncode == OPTCOSINE) {
      for(i = 0; i < n; i++) y[i] = m1optcosine(x[i], mean[i], sd[i]);
    }
  } else if(*m == 2) {
    /* partial second moment */
    if(*kerncode == GAUSSIAN) {
      for(i = 0; i < n; i++) y[i] = m2gaussian(x[i], mean[i], sd[i]);
    } else if(*kerncode == RECTANGULAR){ 
      for(i = 0; i < n; i++) y[i] = m2rectangular(x[i], mean[i], sd[i]);
    } else if(*kerncode == TRIANGULAR) {
      for(i = 0; i < n; i++) y[i] = m2triangular(x[i], mean[i], sd[i]);
    } else if(*kerncode == EPANECHNIKOV) {
      for(i = 0; i < n; i++) y[i] = m2epanechnikov(x[i], mean[i], sd[i]);
    } else if(*kerncode == BIWEIGHT) {     
      for(i = 0; i < n; i++) y[i] = m2biweight(x[i], mean[i], sd[i]);
    } else if(*kerncode == COSINE) {       
      for(i = 0; i < n; i++) y[i] = m2cosine(x[i], mean[i], sd[i]);
    } else if(*kerncode == OPTCOSINE) {
      for(i = 0; i < n; i++) y[i] = m2optcosine(x[i], mean[i], sd[i]);
    }
  } else {
    *errcode = NOT_SUPPORTED;
    return;
  }
}
